File: tests/native/test_trackball_navigation.cpp
#include <cstdlib>
#include <iostream>
#include <vector>

#include "../../lib/tdeck_ui/Hardware/TDeck/TrackballNavigation.h"

using Hardware::TDeck::NavigationCandidate;
using Hardware::TDeck::NavigationDirection;
using Hardware::TDeck::NavigationRect;
using Hardware::TDeck::select_directional_candidate;

static int failures = 0;

#define EXPECT_EQ(actual, expected) do { \
    const auto actual_value = (actual); \
    const auto expected_value = (expected); \
    if (actual_value != expected_value) { \
        std::cerr << __func__ << ": expected " << expected_value \
                  << ", got " << actual_value << "\n"; \
        ++failures; \
    } \
} while (0)

static NavigationCandidate candidate(int id, int x, int y, int width, int height,
                                     bool visible = true) {
    return NavigationCandidate{id, NavigationRect{x, y, width, height}, visible};
}

static void horizontal_movement_stays_on_the_same_row() {
    const NavigationRect focused{10, 10, 30, 20};
    const std::vector<NavigationCandidate> candidates{
        candidate(1, 50, 10, 30, 20),
        candidate(2, 10, 50, 30, 20),
        candidate(3, 100, 50, 30, 20),
    };

    EXPECT_EQ(select_directional_candidate(focused, candidates.data(), candidates.size(),
                                           NavigationDirection::RIGHT), 1);
}

static void vertical_movement_stays_in_the_same_column() {
    const NavigationRect focused{100, 10, 30, 20};
    const std::vector<NavigationCandidate> candidates{
        candidate(1, 30, 50, 30, 20),
        candidate(2, 100, 60, 30, 20),
        candidate(3, 140, 45, 30, 20),
    };

    EXPECT_EQ(select_directional_candidate(focused, candidates.data(), candidates.size(),
                                           NavigationDirection::DOWN), 2);
}

static void every_direction_is_distinct() {
    const NavigationRect focused{100, 100, 20, 20};
    const std::vector<NavigationCandidate> candidates{
        candidate(1, 100, 50, 20, 20),
        candidate(2, 100, 150, 20, 20),
        candidate(3, 50, 100, 20, 20),
        candidate(4, 150, 100, 20, 20),
    };

    EXPECT_EQ(select_directional_candidate(focused, candidates.data(), candidates.size(), NavigationDirection::UP), 1);
    EXPECT_EQ(select_directional_candidate(focused, candidates.data(), candidates.size(), NavigationDirection::DOWN), 2);
    EXPECT_EQ(select_directional_candidate(focused, candidates.data(), candidates.size(), NavigationDirection::LEFT), 3);
    EXPECT_EQ(select_directional_candidate(focused, candidates.data(), candidates.size(), NavigationDirection::RIGHT), 4);
}

static void hidden_and_wrong_direction_candidates_are_ignored() {
    const NavigationRect focused{100, 100, 20, 20};
    const std::vector<NavigationCandidate> candidates{
        candidate(1, 100, 50, 20, 20),
        candidate(2, 100, 150, 20, 20, false),
        candidate(3, 100, 90, 20, 20),
    };

    EXPECT_EQ(select_directional_candidate(focused, candidates.data(), candidates.size(),
                                           NavigationDirection::DOWN), -1);
}

int main() {
    horizontal_movement_stays_on_the_same_row();
    vertical_movement_stays_in_the_same_column();
    every_direction_is_distinct();
    hidden_and_wrong_direction_candidates_are_ignored();

    if (failures != 0) return EXIT_FAILURE;
    std::cout << "4 trackball navigation tests passed\n";
    return EXIT_SUCCESS;
}

File: tests/native/test_trackball_navigation.py
"""Compile and execute portable directional trackball navigation regressions."""

from pathlib import Path
import shutil
import subprocess


ROOT = Path(__file__).resolve().parents[2]


def test_trackball_navigation(tmp_path):
    cxx = shutil.which("clang++") or shutil.which("g++")
    if not cxx:
        raise RuntimeError("A C++17 compiler is required for this test")

    binary = tmp_path / "trackball_navigation"
    result = subprocess.run(
        [
            cxx,
            "-std=c++17",
            "-Wall",
            "-Wextra",
            "-Werror",
            str(ROOT / "tests/native/test_trackball_navigation.cpp"),
            "-o",
            str(binary),
        ],
        cwd=ROOT,
        text=True,
        capture_output=True,
    )
    assert result.returncode == 0, result.stdout + result.stderr

    run = subprocess.run([str(binary)], text=True, capture_output=True)
    assert run.returncode == 0, run.stdout + run.stderr
    assert "4 trackball navigation tests passed" in run.stdout


def test_trackball_and_nomadnet_directional_integration_contract():
    trackball = (ROOT / "lib/tdeck_ui/Hardware/TDeck/Trackball.cpp").read_text()
    nomadnet = (ROOT / "lib/tdeck_ui/UI/LXMF/NomadNetScreen.cpp").read_text()

    assert "LV_KEY_NEXT" not in trackball
    assert "LV_KEY_PREV" not in trackball
    assert "NavigationDirection::UP" in trackball
    assert "NavigationDirection::DOWN" in trackball
    assert "NavigationDirection::LEFT" in trackball
    assert "NavigationDirection::RIGHT" in trackball
    assert "navigate_or_scroll" in trackball
    # Modal dialogs temporarily move the trackball to an isolated group. The
    # callback must navigate that assigned group, never the global default.
    assert "_indev ? _indev->group : nullptr" in trackball
    # Clipped/off-screen controls are not directional candidates. Scrolling
    # reveals them before they can receive focus.
    assert "lv_obj_is_visible(object)" in trackball
    assert "if (!focused || object_is_hidden(focused))" in trackball
    assert "if (!focused || !object_is_visible(focused))" not in trackball
    # Generic scroll fallback is scoped to the assigned group's common UI
    # subtree, so modal input cannot scroll the screen behind it.
    assert "group_navigation_root(group, focused)" in trackball
    assert "find_scroll_target(lv_scr_act()" not in trackball
    assert "if (group->frozen) return false;" in trackball
    assert "!lv_obj_has_state(object, LV_STATE_DISABLED)" in trackball

    # NomadNet's document viewport must remain a bounded vertical scroll target;
    # trackball movement reaches it through the generic LVGL navigation helper.
    assert "lv_obj_set_scroll_dir(_content,LV_DIR_VER)" in nomadnet
    assert "LV_OBJ_FLAG_SCROLLABLE" in nomadnet
